### PR TITLE
Allow specified UDP cmsg to be saved to QuicReceivedPacket

### PR DESCRIPTION
**Pull request by pksohn**
Commit Message: Allow specified UDP cmsg to be saved to QuicReceivedPacket
Additional Description: This can be accessed via QuicListenerFilter::onFirstPacketReceived.
Risk Level: Low
Testing: Integration test
Docs Changes: N/A 
Release Notes: added

### CODE REVIEW

File: source/common/network/io_socket_handle_impl.cc [hunk 1]
```diff
@@ -335,6 +336,11 @@ Api::IoCallUint64Result IoSocketHandleImpl::recvmsg(Buffer::RawSlice* slices,
     // Get overflow, local address and gso_size from control message.
     for (struct cmsghdr* cmsg = CMSG_FIRSTHDR(&hdr); cmsg != nullptr;
          cmsg = CMSG_NXTHDR(&hdr, cmsg)) {
+      if (cmsg->cmsg_type == save_cmsg_config.optname &&
```
Comment (RyanTheOptimist) [inline source/common/network/io_socket_handle_impl.cc:339]: Does this "if" return false when there is no config set? If so how? (I that because the default value for these two fields is 0, and no such cmsg is valid?

File: api/envoy/config/listener/v3/quic_config.proto [hunk 2]
```diff
@@ -24,11 +24,17 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#protodoc-title: QUIC listener config]
 
 // Configuration specific to the UDP QUIC listener.
-// [#next-free-field: 12]
+// [#next-free-field: 13]
 message QuicProtocolOptions {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.api.v2.listener.QuicProtocolOptions";
 
+  message SaveCmsgConfig {
```
Comment (RyanTheOptimist) [inline api/envoy/config/listener/v3/quic_config.proto:32]: Naming is one of the two hard problems in computer science (along with cache invalidation and off by one errors), but I'm not loving "SaveCmsgConfig". I'm also not finding obviously better names. I think that "Save" reads like a verb, so this seems like it's a function name instead of a data structure. Maybe "SavedCmsgConfig"? Oh, or perhaps "CmsReportingConfig" since it configures how CMSG is reported? That seems a bit cleaner to me. (or maybe CmsgSavingConfig if we think "save" is more correct than "report". Or maybe "processing" or "handling"?)

WDYT?

[the remark this answers is not in the document]
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto:32]: Do you think the comment on the new field below suffices?

File: api/envoy/config/listener/v3/quic_config.proto [hunk 3]
```diff
@@ -86,4 +92,7 @@ message QuicProtocolOptions {
   // If not specified, no debug visitor will be attached to connections.
   // [#extension-category: envoy.quic.connection_debug_visitor]
   core.v3.TypedExtensionConfig connection_debug_visitor_config = 11;
+
+  // Configure a type of UDP cmsg to pass to listener filters via QuicReceivedPacket.
```
[the remark this answers is not in the document]
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto:96]: Correct.

File: source/common/network/io_socket_handle_impl.cc (hunk 1 above)
Comment (pksohn) [reply source/common/network/io_socket_handle_impl.cc:339]: Yes, we're relying on the zero defaults. Do you think we should use `optional` and check for presence?

File: api/envoy/config/listener/v3/quic_config.proto (hunk 2 above)
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto:32]: Yup, I don't love the naming either, it was loosely inspired by [TCP_SAVE_SYN](https://lore.kernel.org/all/20150501201417.GB6113@akamai.com/T/).

I think I like "SavedCmsgConfig" best from your alternatives, but I don't love it. Maybe "CmsgToSaveConfig"? Let's settle on an alternative before I change all the callsites.

File: api/envoy/config/listener/v3/quic_config.proto [hunk 4]
```diff
@@ -24,11 +24,17 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
-// [#next-free-field: 12]
+// [#next-free-field: 13]
 message QuicProtocolOptions {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.api.v2.listener.QuicProtocolOptions";
 
+  message SaveCmsgConfig {
+    int32 level = 1;
+
+    int32 optname = 2;
```
[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply api/envoy/config/listener/v3/quic_config.proto:35]: Oh shoot, I missed that. Yeah, I think I would do that since it matches the CMSG API.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply api/envoy/config/listener/v3/quic_config.proto:32]: I think it would be good to explain here what the fields mean. Just a brief description that this config specifies a specific level + type of CMSG data to save, or some such.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply api/envoy/config/listener/v3/quic_config.proto:96]: Since Envoy APIs tend to hang around stone forever, would we every want to save more than one CMSG type? I know we have code that reads the TOS byte to get ECN bits, for example. So maybe we might want to use this mechanism to save this data? Also, for ECN the level is IPPROTO_IP for IPv4 and IPPROTO_IPV6 for IPv6, which might suggest a need for both levels to be specified?

File: source/common/network/io_socket_handle_impl.cc (hunk 1 above)
Comment (RyanTheOptimist) [reply source/common/network/io_socket_handle_impl.cc:339]: I think so. It *probably* doesn't matter too much, but I think using `optional` would be a bit more explicit.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 2 above)
Comment (RyanTheOptimist) [reply api/envoy/config/listener/v3/quic_config.proto:32]: Ah, interesting! I see how that maps to what you wrote. I'm happy to go with "SaveCmsgConfig" (even though I still don't love it :>). Thanks for the explanation.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto:96]: Right now we're taking advantage of the so-far unused `packet_headers` and `headers_length` fields in [QuicReceivedPacket](https://github.com/envoyproxy/envoy/blob/dde04ae6c6c3b5af191fbe16aa3b27d91226bafc/source/common/quic/active_quic_listener.cc#L140C54-L140C61) which currently allows for one header, so it'd require a bigger refactor. I agree it seems like a possible need in the future. I'd say we can come back to that if/when we need? Let me know what you think.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto]: Done!

File: envoy/network/io_handle.h [hunk 5]
```diff
@@ -160,6 +163,11 @@ class IoHandle {
     absl::FixedArray<RecvMsgPerPacketInfo> msg_;
   };
 
+  struct UdpSaveCmsgConfig {
```
[the remark this answers is not in the document]
Comment (pksohn) [reply envoy/network/io_handle.h:167]: Done.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto:34]: Added.

**Review by RyanTheOptimist — APPROVED** [review]
LGTM. Still need API review.

File: api/envoy/config/listener/v3/quic_config.proto [hunk 6]
```diff
@@ -86,4 +94,7 @@ message QuicProtocolOptions {
   // If not specified, no debug visitor will be attached to connections.
   // [#extension-category: envoy.quic.connection_debug_visitor]
   core.v3.TypedExtensionConfig connection_debug_visitor_config = 11;
+
+  // Configure a type of UDP cmsg to pass to listener filters via QuicReceivedPacket.
+  SaveCmsgConfig save_cmsg_config = 12;
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:99]: What will be the default behavior (i.e., if it is not set)?

File: api/envoy/config/listener/v3/quic_config.proto [hunk 7]
```diff
@@ -24,11 +24,19 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#protodoc-title: QUIC listener config]
 
 // Configuration specific to the UDP QUIC listener.
-// [#next-free-field: 12]
+// [#next-free-field: 13]
 message QuicProtocolOptions {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.api.v2.listener.QuicProtocolOptions";
 
+  // Configuration for UDP cmsg to pass to listener filters via QuicReceivedPacket.
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:32]: What is cmsg?
Are there links that can be added here that provide more info?

File: api/envoy/config/listener/v3/quic_config.proto [hunk 8]
```diff
@@ -24,11 +24,19 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
+  // Both level and type must be specified for cmsg to be saved.
+  message SaveCmsgConfig {
+    optional int32 level = 1;
+
+    optional int32 type = 2;
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:37]: Please add a description to each field, including the meaning of the default value (0) in each case. Also, just to make sure, the use of signed integer is intentional, correct?

File: api/envoy/config/listener/v3/quic_config.proto [hunk 9]
```diff
@@ -24,11 +24,19 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
-// [#next-free-field: 12]
+// [#next-free-field: 13]
 message QuicProtocolOptions {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.api.v2.listener.QuicProtocolOptions";
 
+  // Configuration for UDP cmsg to pass to listener filters via QuicReceivedPacket.
+  // Both level and type must be specified for cmsg to be saved.
+  message SaveCmsgConfig {
+    optional int32 level = 1;
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:35]: Please remove `optional` from both fields.
If the default value should be something other than 0, please use a WKT field instead.

**Review by adisuissa — COMMENTED** [review]
Thanks!
I'll take the API review

File: api/envoy/config/listener/v3/quic_config.proto (hunk 7 above)
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto:32]: See https://man7.org/linux/man-pages/man3/cmsg.3.html, basically socket control info. I've added a link in the proto.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 6 above)
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto:102]: No cmsg will be saved to the QuicReceivedPacket. I've added a comment.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 8 above)
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto]: Descriptions added. The default value is now unset. Yes, I used a signed integer even though it seems like negatives are never used, because cmsg_level and cmsg_type in the underlying data structure are signed: https://man7.org/linux/man-pages/man3/cmsg.3.html.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 9 above)
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto]: Done, updated to the Int32Value wrapper.

File: source/common/network/io_socket_handle_impl.cc [hunk 10]
```diff
@@ -283,6 +283,7 @@ absl::optional<uint32_t> maybeGetPacketsDroppedFromHeader([[maybe_unused]] const
 
 Api::IoCallUint64Result IoSocketHandleImpl::recvmsg(Buffer::RawSlice* slices,
                                                     const uint64_t num_slice, uint32_t self_port,
+                                                    const UdpSaveCmsgConfig& save_cmsg_config,
```
Comment (danzh2010) [inline source/common/network/io_socket_handle_impl.cc:286]: You may need to allocate cbuf at line290 with larger size, ideally, the config knob should specify the expected type size.

File: envoy/network/io_handle.h [hunk 11]
```diff
@@ -172,7 +183,9 @@ class IoHandle {
    * err_ = nullptr and rc_ = the bytes received for success.
    */
   virtual Api::IoCallUint64Result recvmsg(Buffer::RawSlice* slices, const uint64_t num_slice,
-                                          uint32_t self_port, RecvMsgOutput& output) PURE;
+                                          uint32_t self_port,
+                                          const UdpSaveCmsgConfig& save_cmsg_config,
```
Comment (danzh2010) [inline envoy/network/io_handle.h:187]: please add comment about the new param.

File: source/common/network/io_socket_handle_impl.cc (hunk 10 above)
Comment (pksohn) [reply source/common/network/io_socket_handle_impl.cc:286]: Thanks a lot for this catch. I added a field in the config proto. cc @adisuissa @RyanTheOptimist

File: envoy/network/io_handle.h (hunk 11 above)
Comment (pksohn) [reply envoy/network/io_handle.h:189]: Done.

File: envoy/network/listener.h [hunk 12]
```diff
@@ -446,6 +447,8 @@ class UdpListenerCallbacks {
    * An estimated number of UDP packets this callback expects to process in current read event.
    */
   virtual size_t numPacketsExpectedPerEventLoop() const PURE;
+
+  virtual const IoHandle::UdpSaveCmsgConfig& udpSaveCmsgConfig() const PURE;
```
Comment (danzh2010) [inline envoy/network/listener.h:451]: comment please

File: changelogs/current.yaml [hunk 13]
```diff
@@ -24,5 +24,10 @@ removed_config_or_runtime:
     path.
 
 new_features:
+- area: quic
+  change: |
+    Added QUIC protocol option :ref:save_cmsg_config`
+    <envoy_v3_api_field_config.listener.v3.QuicProtocolOptions.save_cmsg_config>` to allow a configurable CMSG type to be saved to
+    QuicReceivedPacket, which can be accessed via QuicListenerFilter::onFirstPacketReceived.
```
Comment (danzh2010) [inline changelogs/current.yaml:31]: This is very much about implementation details. Can you rephrase it to be more readable to users who doesn't know what QuicReceivedPacket and onFirstPacketReceived are, i.e. "Added ... to propagate the configured CMSG header of the first packet on the connection to QuicListenerFilter."

Comment (pksohn) [reply changelogs/current.yaml]: Thanks, I added a slightly modified version of your suggestion.

**Review by danzh2010 — APPROVED** [review]

**Review by adisuissa — COMMENTED** [review]
High-level API observation and question:
The CMSG seems to be a very low-level detail of an OS-networking capability.
If I understand correctly, the idea is to attach the CMSG data into the `quic::QuicReceivedPacket` data-structure. I'm not sure if adding that ability is core to the QUIC based listener.
Would it make more sense to have the CMSG as an extension (similar to other fields, such as `connection_id_generator_config`), and then one of the implementations could be the CMSG one?
This way there's no coupling on the need CMSG in order to use QUIC.

**Comment by danzh2010** [comment]
> High-level API observation and question: The CMSG seems to be a very low-level detail of an OS-networking capability. If I understand correctly, the idea is to attach the CMSG data into the `quic::QuicReceivedPacket` data-structure. I'm not sure if adding that ability is core to the QUIC based listener.

This can be used for any UDP listener which share the same socket read code.

> Would it make more sense to have the CMSG as an extension (similar to other fields, such as `connection_id_generator_config`), and then one of the implementations could be the CMSG one? This way there's no coupling on the need CMSG in order to use QUIC.

Unless the whole packet reading logic in IoHandle and listener callback is moved into an extension, I don't see how to define a clean extension API.

File: api/envoy/config/listener/v3/quic_config.proto [hunk 14]
```diff
@@ -24,11 +24,25 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
-// [#next-free-field: 12]
+// [#next-free-field: 13]
 message QuicProtocolOptions {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.api.v2.listener.QuicProtocolOptions";
 
+  // Configuration for UDP cmsg to pass to listener filters via QuicReceivedPacket.
+  // Both level and type must be specified for cmsg to be saved.
+  // More info about CMSG: https://man7.org/linux/man-pages/man3/cmsg.3.html
+  message SaveCmsgConfig {
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:35]: This is probably more of a core type (may be used not only for QUIC, but more generic to a socket), so I think the following should be done:
1. it should be moved to a new package in: api/envoy/config/core/v3/.
2. it should probably be renamed to something like `SocketCmsgHeader` as it reflects the structure of a CMSG header (as defined in https://man7.org/linux/man-pages/man3/cmsg.3.html), IIUC.

File: api/envoy/config/listener/v3/quic_config.proto [hunk 15]
```diff
@@ -86,4 +100,8 @@ message QuicProtocolOptions {
   // If not specified, no debug visitor will be attached to connections.
   // [#extension-category: envoy.quic.connection_debug_visitor]
   core.v3.TypedExtensionConfig connection_debug_visitor_config = 11;
+
+  // Configure a type of UDP cmsg to pass to listener filters via QuicReceivedPacket.
+  // If not specified, no cmsg will be saved to QuicReceivedPacket.
+  SaveCmsgConfig save_cmsg_config = 12;
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:106]: This should probably be defined as a repeated field (adda a validator ensuring that only 1 element appears if the code only supports that - if this won't be defined as repeated now and needed in the future it will require a new field).

File: api/envoy/config/listener/v3/quic_config.proto [hunk 16]
```diff
@@ -24,11 +24,25 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
 // Configuration specific to the UDP QUIC listener.
-// [#next-free-field: 12]
+// [#next-free-field: 13]
 message QuicProtocolOptions {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.api.v2.listener.QuicProtocolOptions";
 
+  // Configuration for UDP cmsg to pass to listener filters via QuicReceivedPacket.
+  // Both level and type must be specified for cmsg to be saved.
+  // More info about CMSG: https://man7.org/linux/man-pages/man3/cmsg.3.html
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:34]:
```suggestion
  // More info about `:ref:CMSG <https://man7.org/linux/man-pages/man3/cmsg.3.html>`_.
```

File: api/envoy/config/listener/v3/quic_config.proto [hunk 17]
```diff
@@ -24,11 +24,25 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
+  message SaveCmsgConfig {
+    // cmsg level to save if matched. Default is unset and no cmsg will be saved.
+    google.protobuf.Int32Value level = 1;
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:37]: Looking a bit closer at CMSG, it seems that the type for `level`, `type` and `expected_size` should be unsigned.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 14 above)
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto:35]: Thanks for the feedback. I moved this and renamed this as suggested, with some tweaks to the comments to make them more generalized.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 15 above)
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto:106]: Done, with validation.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 17 above)
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto:37]: Done; this requires a static cast to int for equality check with the [cmsghdr int types](https://man7.org/linux/man-pages/man3/cmsg.3.html).

File: api/envoy/config/listener/v3/quic_config.proto [hunk 18]
```diff
@@ -86,4 +87,11 @@ message QuicProtocolOptions {
   // If not specified, no debug visitor will be attached to connections.
   // [#extension-category: envoy.quic.connection_debug_visitor]
   core.v3.TypedExtensionConfig connection_debug_visitor_config = 11;
+
+  // Configure a type of UDP cmsg to pass to listener filters via QuicReceivedPacket.
+  // Both level and type must be specified for cmsg to be saved.
+  // Cmsg may be truncated or omitted if expected size is not set.
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:93]: I wonder if this part of the comment true for CMSG headers in general (and then it should probably be part of the description of `SocketCmsgHeaders`. However, if it is only related to the QUIC related implementation, then I agree it should be specified here.

File: api/envoy/config/core/v3/socket_cmsg_headers.proto [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+option java_package = "io.envoyproxy.envoy.config.core.v3";
+option java_outer_classname = "SocketCmsgHeadersProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/config/core/v3;corev3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Socket CMSG headers]
+
+// Configuration for socket cmsg headers.
+// More info about `:ref:CMSG <https://man7.org/linux/man-pages/man3/cmsg.3.html>`_.
```
Comment (adisuissa) [inline api/envoy/config/core/v3/socket_cmsg_headers.proto:18]: nit:
```suggestion
// See `:ref:CMSG man page<https://man7.org/linux/man-pages/man3/cmsg.3.html>`_ for further information.
```

**Review by adisuissa — COMMENTED** [review]
Overall API LGTM. Left minor comment/question about the comments.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 18 above)
Comment (pksohn) [reply api/envoy/config/listener/v3/quic_config.proto:93]: As far as I know, this is specific to QUIC, but that's also because I simply don't know what other implementations might look like.

File: api/envoy/config/core/v3/socket_cmsg_headers.proto (hunk 19 above)
Comment (pksohn) [reply api/envoy/config/core/v3/socket_cmsg_headers.proto]: Thanks, done.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Review by danzh2010 — APPROVED** [review]

**Review by RyanTheOptimist — DISMISSED** [review]

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Review by RyanTheOptimist — APPROVED** [review]
